rspace/src/test/scala/coop/rchain/rspace/history/HistoryActionsTests.scala

```diff
@@ -128,14 +129,273 @@ abstract class HistoryActionsTests[T] extends HistoryTestsBase[T, TestKey, ByteV
+      lookup(store, key2) shouldBe None
+    }
+
+  "Insert, delete, delete again" should "work" in
+    withTestTrieStore { (store: ITrieStore[T, TestKey, ByteVector]) =>
+      val root0 = store.workingRootHash.get
+
+      insert(store, key1, val1)
+      lookup(store, key1).value shouldBe val1
+      store.workingRootHash.get should not be root0
+
+      delete(store, key1, val1) shouldBe true
+      lookup(store, key1) shouldBe None
+      store.workingRootHash.get shouldBe root0
+
+      delete(store, key1, val1) shouldBe false
+      lookup(store, key1) shouldBe None
+      store.workingRootHash.get shouldBe root0
+    }
+
+  "Insert, Insert, Delete, Delete, Rollback, Delete, Delete" should "work" in
+    withTestTrieStore { (store: ITrieStore[T, TestKey, ByteVector]) =>
+      val root0 = store.workingRootHash.get
+
+      insert(store, key1, val1)
+      val root1 = store.workingRootHash.get
+      root1 should not be root0
+
+      insert(store, key2, val2)
+      val root2 = store.workingRootHash.get
+      root2 should not be root1
+
+      lookup(store, key1).value shouldBe val1
+      lookup(store, key2).value shouldBe val2
+
+      delete(store, key2, val2) shouldBe true
+      store.workingRootHash.get shouldBe root1
+
+      lookup(store, key1).value shouldBe val1
+      lookup(store, key2) shouldBe None
+
+      delete(store, key1, val1) shouldBe true
+      store.workingRootHash.get shouldBe root0
+
+      lookup(store, key1) shouldBe None
+      lookup(store, key2) shouldBe None
+
+      // Aaannnd rollback...
+      store.workingRootHash.take()
+      store.workingRootHash.put(root2)
+
+      lookup(store, key1).value shouldBe val1
+      lookup(store, key2).value shouldBe val2
+
+      delete(store, key2, val2) shouldBe true
+      store.workingRootHash.get shouldBe root1
+
+      lookup(store, key1).value shouldBe val1
+      lookup(store, key2) shouldBe None
+
+      delete(store, key1, val1) shouldBe true
+      store.workingRootHash.get shouldBe root0
+
+      lookup(store, key1) shouldBe None
+      lookup(store, key2) shouldBe None
+    }
+
+  "Rollback twice" should "work" in
+    withTestTrieStore { (store: ITrieStore[T, TestKey, ByteVector]) =>
+      val root0 = store.workingRootHash.get
+
+      insert(store, key1, val1)
+      val root1 = store.workingRootHash.get
+      root1 should not be root0
+
+      insert(store, key2, val2)
+      val root2 = store.workingRootHash.get
+      root2 should not be root1
+
+      lookup(store, key1).value shouldBe val1
+      lookup(store, key2).value shouldBe val2
+
+      delete(store, key2, val2) shouldBe true
+      store.workingRootHash.get shouldBe root1
+
+      lookup(store, key1).value shouldBe val1
+      lookup(store, key2) shouldBe None
+
+      delete(store, key1, val1) shouldBe true
+      store.workingRootHash.get shouldBe root0
+
+      lookup(store, key1) shouldBe None
+      lookup(store, key2) shouldBe None
+
+      // Aaannnd rollback...
```

Comment:
I think it would make sense to have a `rollbackTo(hash: Blake2b256Hash)` method defined for the `store` encapsulating the 2 operations below to avoid the possibility of falling into an infinite loop in `put`

Reply:
Yes, 100%.  Will do

Reply:
Done.